### ISSUE TITLE
feat: GNN node classification pipeline for code smell and vulnerability detection (#779)

### ISSUE THREAD

**Issue by danielperezr88** [danielperezr88/astrolabe#809]
## Summary

Build a GNN node classification pipeline that enables training Graph Neural Networks on Astrolabe code graphs for automated code smell detection, vulnerability detection, and maintainability prediction. This is the ML infrastructure layer that consumes the Python Dataset API (issue #804).

## Research Basis

- **Long Method Detection via GCN (Liu et al., 2023)**: Extended PDG to Directed-Heterogeneous Graph + 2-layer GCN. Uses tree-sitter for AST, 14 software metrics as node features. 96.3% accuracy on Long Method.
- **HMML Multi-Label Code Smell (SEKE 2022)**: CPG (AST+CFG+PDG) + GCN + BiLSTM with attention. Multi-label classification. F1: 0.78 macro avg.
- **Vul-LMGNN (2024)**: CodeBERT node embeddings + gated GGNN on CPG. Online knowledge distillation. Outperforms 17 SOTA methods.
- **Devign (NeurIPS 2019)**: GGNN + 1D Conv for graph-level vulnerability classification. The pioneer GNN-for-code tool.
- **Integrated GNN for Defect + Quality (Nature Sci Rep, 2025)**: Dual-branch GNN on multi-level graph. F1: 0.811, AUC: 0.896.

### 1. Node Feature Engineering Pipeline
- Type encoding: one-hot for 49 node labels
- Structural features: degree (in/out), PageRank score, betweenness centrality, community membership
- Code metrics: parameterCount, visibility encoding, isAsync/isStatic/isAbstract flags
- Compositional features: lines of code (endLine - startLine), nesting level
- Optional: code embedding from Transformers.js (384D vector from all-MiniLM-L6-v2)

### 2. Edge Feature Encoding
- Relationship type: one-hot for 26 edge types
- Confidence score: scalar 0-1
- Step ordinal (for STEP_IN_PROCESS): normalized scalar
- Cross-community flag: binary

### 3. Heterogeneous GNN Model Architectures
- **HGTConv**: Native heterogeneous transformer — handles all node/edge types
- **HeteroConv**: Per-edge-type custom operators (GATConv for CALLS, SAGEConv for IMPORTS)
- **to_hetero()**: Wrap homogeneous GCN for quick prototyping
- All models output node-level or graph-level predictions

### 4. Pre-built Training Pipelines
- **Code Smell Detection**: Binary/multi-label node classification (God Class, Long Method, Feature Envy)
- **Vulnerability Detection**: Binary node classification (vulnerable/safe)
- **Maintainability Prediction**: Regression on node-level maintainability score
- **Architecture Role Classification**: Predict architectural role (controller, service, model, utility) from graph topology

### 5. Training Infrastructure
- PyTorch Lightning integration for reproducible training
- Cross-validation with repository-level splits (avoid data leakage)
- Hyperparameter search via Optuna
- Model evaluation: accuracy, F1, AUC, confusion matrix per node type
- Export trained models for inference via MCP tool

### 6. Inference via MCP
After training, expose model predictions through a new MCP tool:
```
predict(model="code_smell_v1", repo="...", nodes=["AuthService.login"])
  -> { node: "AuthService.login", prediction: "long_method", confidence: 0.87 }
```

## Labels

- **Category**: gnn-ml, graph-analytics
- **Effort**: large (1-2 months)
- **Risk**: medium-risk (well-understood GNN architectures, but labeling data requires effort)
- **Benefit**: Transforms Astrolabe from analysis tool to ML-powered code intelligence platform
- **Depends on**: Issue #804 (Python Graph Dataset API)

**Comment by danielperezr88** [comment]
Fixed in PR #824 (merged).
